src/lib/SaferNames/Syntax.hs

```diff
@@ -436,147 +355,509 @@ instance Semigroup (SourceNameMap n) where
 
 instance HasNamesB DataConRefBinding where
-  traverseNamesB = undefined
-  boundScope = undefined
-
-instance SubstB b => SubstB (Nest b) where
-  traverseCoreB s t nest = case nest of
-    Empty -> do return $ FreshBinder emptyFreshExt Empty emptyEnv
-    Nest b rest ->
-      traverseCoreB s t (NestPair b rest) >>= \case
-        FreshBinder ext (NestPair b' rest') renamer ->
-          return $ FreshBinder ext (Nest b' rest') renamer
-  asBindingsFrag = undefined
-
-instance (SubstB b1, SubstB b2) => SubstB (NestPair b1 b2) where
-  traverseCoreB = undefined
-  asBindingsFrag = undefined
-
-instance HasNamesE NamedDataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE NamedDataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE DataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE DataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE Atom where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Atom where
-  traverseCoreE s t atom = case atom of
-    Var (AnnVar v ann) ->
-      lookupNameTraversal t v >>= \case
-        Left substVal -> case substVal of
-          SubstVal x -> return x
-          Rename v' -> do
-            ann' <- traverseCoreE s t ann
-            return $ Var $ AnnVar v' ann'
-        Right v' -> do
-          ann' <- traverseCoreE s t ann
-          return $ Var $ AnnVar v' ann'
-    Lam ab -> Lam <$> tce s t ab
-    Pi  ab -> Pi  <$> tce s t ab
-    DataCon def params con args ->
-      DataCon <$> tce s t def <*> mapM (tce s t) params
-              <*> pure con <*> mapM (tce s t) args
-    TypeCon def params -> TypeCon <$> tce s t def <*> mapM (tce s t) params
-    -- LabeledRow (ExtLabeledItems (Type n) (Name n))
-    Record items -> Record <$> mapM (tce s t) items
-    -- RecordTy  (ExtLabeledItems (Type n) (Name n))
-    -- Variant   (ExtLabeledItems (Type n) (Name n)) Label Int (Atom n)
-    -- VariantTy (ExtLabeledItems (Type n) (Name n))
-    Con con -> Con <$> traverse (tce s t) con
-    TC  con -> TC  <$> traverse (tce s t) con
-    Eff effs -> Eff <$> tce s t effs
-    ACase scrut alts ty ->
-      ACase <$> tce s t scrut <*> traverse (tce s t) alts <*> tce s t ty
-    -- DataConRef (DataDef n) [Atom n] (EmptyNest DataConRefBinding n)
-    -- BoxedRef (Atom n) (Atom n) (Abs Binder Block n)  -- ptr, size, binder/body
-    -- ProjectElt (NE.NonEmpty Int) (Var n)
-    _ -> undefined
-
-instance HasNamesE Expr where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Expr where
-  traverseCoreE s t expr = case expr of
-    App e1 e2 -> App <$> tce s t e1 <*> tce s t e2
+  traverseNamesB (DataConRefBinding b ref) cont = do
+    ref' <- traverseNamesE ref
+    traverseNamesB b \b' -> cont $ DataConRefBinding b' ref'
+
+instance SubstB AtomSubstVal DataConRefBinding where
+  substB (DataConRefBinding b ref) cont = do
+    ref' <- substE ref
+    substB b \b' -> cont $ DataConRefBinding b' ref'
+
+instance InjectableE Atom where
```

Comment:
I'd still like to better understand the utility of those `InjectableE` instances, but it shouldn't block this